Background task to check for any repos w/ last_error nonempty and update

mollylogue opened the pull request:
background task in repo-updater to check for repos w/ nonempty last_error and update

Fixes https://github.com/sourcegraph/sourcegraph/issues/25217

TODOs:
- Finalize time interval for sleep
- Rebase once https://github.com/sourcegraph/sourcegraph/pull/28655 is merged and update to use the synchronous version of `syncer.SyncRepo`, then remove sleep statement from test.

Questions:
- How often does the repo-updater main function get run? My assumption is that this task will kick off on that cadence and fetch the list of all the error-state repos that need to be updated and update them in the background, but I want to make sure this assumption is correct.
- Is it okay to use the same `Syncer` for this task as is used everywhere else? I ask because presumably then all these tasks are publishing to the same `Synced` [channel](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/repos/syncer.go?L34) and I'm not sure if that's okay or not.
- Should I be concerned that there are errors in the logs when the (now turned private) repo is synced, since it can no longer be found? They don't cause the task to exit or anything, just add extra errors to our logs. 


NOTE: I'm adding a super rough screen recording of the behavior with this code. You'll see that I see the repo when it's public, then when I turn it to private and navigate back to it on sourcegraph.com, I still see it. When I check the logs I see that `gitserver` has run into an error (since it can't find the repo anymore). Then, when I restart to trigger the repo-updater running, now the repo updater finds these gitserver errors in the `gitserver_repos` table and updates the repo, causing it to be deleted. When I navigate back to the page, you see that now it displays a `404 Repository Not Found` as desired. 


UPDATE: This is now split into two PRs. https://github.com/sourcegraph/sourcegraph/pull/28554 adds in some changes around the `repo.Store` code. This PR is based off of that branch for now for ease of review/commenting but should be rebased on main once the other PR is merged. 

UPDATE: (12/8/21) I'd like to get https://github.com/sourcegraph/sourcegraph/pull/28655 merged first so that I can rebase and improve the testing of this by removing that asynchronous logic.

<!-- Reminder: Have you updated the changelog and relevant docs (user docs, architecture diagram, etc) ? -->
<!-- Please notify @delivery if this PR contains changes to CI that may need to be cherry-picked on to patch release branches -->

mollylogue commented:
Okay, my screen recording is not able to be uploaded for some very sad unknown reason. Hopefully I can get that up tomorrow or just post in slack somewhere instead.

ryanslade commented:
> * How often does the repo-updater main function get run? My assumption is that this task will kick off on that cadence and fetch the list of all the error-state repos that need to be updated and update them in the background, but I want to make sure this assumption is correct.

It's only run once on startup so I think we should instead launch a background goroutine that does the following: 

1. Fetches the list of problematic repos
1. For each repo
  1. Wait for rate limiter
  2. Sync
1. Once we've done all repos, sleep for some period of time and loop back around  

> * Is it okay to use the same `Syncer` for this task as is used everywhere else? I ask because presumably then all these tasks are publishing to the same `Synced` [channel](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/repos/syncer.go?L34) and I'm not sure if that's okay or not.

No problem to use the existing syncer, you could pass it in to the background worker routine mentioned above.

> * Should I be concerned that there are errors in the logs when the (now turned private) repo is synced, since it can no longer be found? They don't cause the task to exit or anything, just add extra errors to our logs.

I think that's ok, we expect errors there as we can no longer access the repo which is what causes us to drop it.

One final note, it might be easier to review if you push a PR that just does all the gitserver store related changes as that's most boilerplate. We can review that separately and then review the smaller PR that adds the background task.

mollylogue commented:
UPDATE: This is now split into two PRs. #28554 adds in some changes around the repo.Store code. This PR is based off of that branch for now for ease of review/commenting but should be rebased on main once the other PR is merged.

I've finally gotten some tests to work, though I'm not satisfied with them just yet. I'd like to get some more scenarios covered if possible. 

I have no idea if I did the rate limiting correctly yet. I was trying to reuse already existing code (which I saw was used in the permissions syncer) rather than write something from scratch, but I still have to write a test scenario to validate that I'm using it correctly.

I've tested this most recent version locally and I can validate that it works as we want for both public -> private visibility change and a deleted repo. 

@ryanslade @indradhanush I'd love some eyes on this whenever y'all have the chance to make sure I'm on the right track, though I know I still have some things to work on and will still be pushing commits.

eseliger commented on internal/database/gitserver_repos.go:
```diff
@@ -82,6 +83,45 @@ INSERT INTO
+	if err != nil {
+		return errors.Wrap(err, "fetching repos with nonempty last_error")
+	}
+	defer rows.Close()
```
we have [basestore.CloseRows](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/repos.go?L1190:1) that also does the error handling. It requires a named error argument :)

ryanslade commented on internal/repos/sync_errored.go:
```diff
@@ -0,0 +1,63 @@
+	return s.Store.GitserverReposStore.IterateWithNonemptyLastError(ctx, func(repo types.RepoGitserverStatus) error {
+		codehost := extsvc.CodeHostOf(repo.Name, extsvc.PublicCodeHosts...)
+
+		err := waitForRateLimit(ctx, rateLimiterRegistry, codehost.ServiceID, n)
```
I don't think we should be using the rate limit registry as this uses our internal rate limiter which isn't enabled on .com AFAIK. 

I think for this code we're fine to just pick a low rate limit, maybe 200 per hour and just pass that in from main using `rate.NewLimiter`. Then all we need to do here is call `rate.Wait()`

mollylogue replied:
Okay great! I honestly was very unsure about how to do this bit and didn't want to re-implement too much so I dug through the codebase for examples and this was the main thing I found. But that sounds like a way better way of doing things 😅

mollylogue replied:
Also, could you point me to any docs about the rate limit registry/internal rate limiter (if we have them)? That's probably a good thing for me to know about.

ryanslade replied:
We have a little bit about internal rate limits here:
https://docs.sourcegraph.com/admin/external_service/github#internal-rate-limits

To be honest there's a lot we could do to improve them and in most cases they just make the code more complicated than they need to be..

ryanslade commented on internal/repos/sync_errored.go:
```diff
@@ -0,0 +1,63 @@
+// Dot com mode.
+func (s *Syncer) SyncReposWithLastErrors(ctx context.Context, rateLimiterRegistry *ratelimit.Registry, n int) error {
+	return s.Store.GitserverReposStore.IterateWithNonemptyLastError(ctx, func(repo types.RepoGitserverStatus) error {
+		codehost := extsvc.CodeHostOf(repo.Name, extsvc.PublicCodeHosts...)
```
`codehost` may come back as `nil`. But, I don't think we need this anyway, see the next comment :)

ryanslade reviewed (commented):
Looks good, just a few minor changes and then we should be good to go

mollylogue commented on internal/database/gitserver_repos_test.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"testing"
 
+	"github.com/sourcegraph/sourcegraph/internal/database/dbutil"
```
Yeah, I was super confused as to why GoLand was grouping the imports this way. I figured you'd tell me what was correct 🤷‍♀️ 😂

mollylogue commented on internal/repos/sync_errored.go:
```diff
@@ -0,0 +1,63 @@
+// table, indicating there was an issue updating the repo, and syncs each of these repos. Repos which are no longer
+// visible (i.e. deleted or made private) will be deleted from the DB. Note that this is only being run in Sourcegraph
+// Dot com mode.
+func (s *Syncer) SyncReposWithLastErrors(ctx context.Context, rateLimiterRegistry *ratelimit.Registry, n int) error {
```
Ah, good catch. So, I know that the `syncer.SyncRepo` function will only sync `cloud_default` repos, but I should definitely also check that field during the first db query to fetch the repos w/ last_error. Otherwise we will be iterating through a lot of repos that we don't need to.

ryanslade replied:
Yep, the best place to do the filtering is in the initial database query

mollylogue commented:
@ryanslade I think I've addressed all comments at this point. The only thing I haven't yet done is decided on a reasonable interval between runs. Any thoughts on that?

ryanslade commented on cmd/repo-updater/shared/main.go:
```diff
@@ -198,6 +200,11 @@ func Main(enterpriseInit EnterpriseInit) {
 	go syncScheduler(ctx, scheduler, store)
 
+	if envvar.SourcegraphDotComMode() {
+		rateLimiter := rate.NewLimiter(200, 1)
```
This is 200 per second :) I think we want it to be a lot lower, maybe 200 per hour so that would be `0.05`.

mollylogue replied:
🤦‍♀️ meant to ask about/update this one too. but that sounds good 👍

ryanslade commented on internal/database/gitserver_repos_test.go:
```diff
@@ -2,14 +2,23 @@ package database
+	"github.com/sourcegraph/sourcegraph/internal/types/typestest"
+
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
```
GoLand doing the wrong thing again.. It does this for me too, would be nice if we could find a solution to this

mollylogue replied:
Yeah, I ended up having to group the imports myself, delete all the spaces and then run goimports 🤷‍♀️ kinda weird.

coury-clark replied:
You should be able to solve this in Goland by setting up file watchers:
<img width="238" alt="image" src="https://user-images.githubusercontent.com/5090588/146054467-92435389-968b-47ff-b640-752a6333c816.png">

ryanslade commented on internal/database/repos.go:
```diff
@@ -61,6 +61,11 @@ func (e *RepoNotFoundErr) NotFound() bool {
 }
 
+func IsRepoNotFoundErr(err error) bool {
+	_, ok := err.(*RepoNotFoundErr)
```
I think it's safer to use `Is` here: https://pkg.go.dev/errors#Is

ryanslade commented on internal/repos/sync_errored.go:
```diff
@@ -0,0 +1,46 @@
+		}
+		_, err = s.SyncRepo(ctx, repo.Name, false)
+		if err != nil && !database.IsRepoNotFoundErr(err) {
+			return err
```
I think we can should just log the error and continue. We may have a problematic repo that blocks us from syncing all the others.

ryanslade commented on internal/repos/syncer.go:
```diff
@@ -358,6 +358,7 @@ func (s *Syncer) syncRepo(
 						"error", err2,
 					)
 				}
+				s.notifyDeleted(ctx, stored.ID)
```
Good spot!

ryanslade commented on internal/repos/sync_errored.go:
```diff
@@ -0,0 +1,46 @@
+	"github.com/sourcegraph/sourcegraph/internal/types"
+)
+
+const syncInterval = 2 * time.Minute // TODO: decide on appropriate interval. Currently set low for ease of testing
```
Since we have a rate limiter anyway it's not too important what we pick here. Let's just go with 5 minutes.

mollylogue replied:
👍

ryanslade commented:
> @ryanslade I think I've addressed all comments at this point. The only thing I haven't yet done is decided on a reasonable interval between runs. Any thoughts on that?

Looks good to me apart from a few minor comments, most importantly the one about he chosen rate limit. 

One final thought is that it would be nice add a gauge counting the number of repos we iterate over on each run as in theory we should see that move down over time.

Thanks for all the work on this so far. There's always a lot of back and forth on the first few PR's :)

mollylogue commented:
@ryanslade Do you think we should add that counter as a part of this PR? I didn't get to it today but could add that to this tomorrow or do a follow-up.

ryanslade commented:
> @ryanslade Do you think we should add that counter as a part of this PR? I didn't get to it today but could add that to this tomorrow or do a follow-up.

On the one hand I don't want this PR to grow too much bigger, but on the other hand it would be nice to have the instrumentation in place from the start as I imagine we'll see a big drop in the first few iterations. 

It should be a fairly small change. We just need to create the gauge and then set it after each iteration. If we fail to wait on the rate limiter we should probably *not* set the value so that we only count a "full" iteration.

indradhanush commented:
> It should be a fairly small change. We just need to create the gauge and then set it after each iteration. If we fail to wait on the rate limiter we should probably not set the value so that we only count a "full" iteration.

Agreed. @mollylogue, you could also stack a new PR on top of this based off this branch and then when this one is merged, GitHub will automatically update it to be against main and we can merge them back to back. This might help move the PRs faster through review since this one has been in the works for some time now. :slightly_smiling_face:

ryanslade commented:
@indradhanush But if we merge this PR then the change will get deployed without visibility via the instrumentation. I'm usually in favour of lots of small incremental PR's but in this case I think it's useful to have the instrumentation go out with the code change.

indradhanush commented:
> @indradhanush But if we merge this PR then the change will get deployed without visibility via the instrumentation. I'm usually in favour of lots of small incremental PR's but in this case I think it's useful to have the instrumentation go out with the code change.

Oh. I was thinking to merge them back to back immediately after both the PRs are reviewed and ready to go. But if it's a small change we could go ahead with this PR as well. :ok_hand:

mollylogue commented:
@indradhanush @ryanslade I think it's best that the instrumentation goes out with these changes, too. I'll work on that today.

ryanslade commented:
Meta: We usually try not to do force pushes since it can make it harder to review what has changed since you last reviewed.

mollylogue commented:
> Meta: We usually try not to do force pushes since it can make it harder to review what has changed since you last reviewed.

Ah sorry about that. It's a force of habit from my last job.

ryanslade commented on internal/repos/sync_errored.go:
```diff
@@ -13,6 +16,11 @@ import (
 
 const syncInterval = 5 * time.Minute
 
+var counter = promauto.NewCounter(prometheus.CounterOpts{
```
I think this should be a gauge since we want to simply track the number of repos with sync errors which will go up and down over time.

Counters are more useful for tracking rates of change as they reset to zero each time a service restarts.

mollylogue replied:
🤦‍♀️ good catch.

ryanslade commented on internal/repos/sync_errored.go:
```diff
@@ -0,0 +1,54 @@
+		if err != nil {
+			log15.Error("error syncing repo", "repo", repo.Name, "err", err)
+		}
+		gauge.Add(1)
```
I think instead of adding one each time we should count the total number of repos we iterate over and then set that final count at the end. If we always add one then the gauge value will increase forever. What we want to see is the current approximate number of "failed" repos.

ryanslade commented on internal/repos/sync_errored.go:
```diff
@@ -0,0 +1,57 @@
+
+const syncInterval = 5 * time.Minute
+
+var gauge = promauto.NewGauge(prometheus.GaugeOpts{
```
Sorry, one more nit 😄 

Since this is a package level variable it should probably have a less generic name. Maybe `problematicRepoGauge`?

mollylogue replied:
Ah good call
